src/mlpack/methods/ann/layer/weight_norm.hpp

```diff
@@ -0,0 +1,228 @@
+ *
+ * For more information, refer to the following paper,
+ *
+ * @code
+ * @inproceedings{Salimans2016WeightNorm,
+ *   title = {Weight Normalization: A Simple Reparameterization to Accelerate
+ *            Training of Deep Neural Networks},
+ *   author = {Tim Salimans, Diederik P. Kingma},
+ *   booktitle = {Neural Information Processing Systems 2016},
+ *   year = {2016}
+ * }
+ * @endcode
+ *
+ * @tparam InputDataType Type of the input data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam OutputDataType Type of the output data (arma::colvec, arma::mat,
+ *         arma::sp_mat or arma::cube).
+ * @tparam CustomLayers Additional custom layers that can be added.
+ */
+template <
+  typename InputDataType = arma::mat,
+  typename OutputDataType = arma::mat,
+  typename... CustomLayers
+>
+class WeightNorm
+{
+ public:
+  //! Create the WeightNorm layer object.
+  WeightNorm();
+
+  //! Destructor to release allocated memory.
+  ~WeightNorm();
+
+  /**
+   * Reset the layer parameters.
+   */
+  void Reset();
+
+  /**
+   * Forward pass of the Weight Normalization layer. Calculates the weights of
+   * the wrapped layer from the parameter vector v and the scalar parameter g.
+   * It then calulates the output of the wrapped layer from the calculated
+   * weights.
+   *
+   * @param input Input data for the layer.
+   * @param output Resulting output activations.
+   */
+  template<typename eT>
+  void Forward(arma::Mat<eT>&& input, arma::Mat<eT>&& output);
+
+  /**
+   * Backward pass through the layer. This function calls the Backward()
+   * function of the wrapped layer.
+   *
+   * @param input The input activations.
+   * @param gy The backpropagated error.
+   * @param g The calculated gradient.
+   */
+  template<typename eT>
+  void Backward(const arma::Mat<eT>&& input,
+                arma::Mat<eT>&& gy,
+                arma::Mat<eT>&& g);
+
+  /**
+   * Calculate the gradient using the output delta and the input activations and
+   * weights of the wrapped layer.
+   *
+   * @param input The input activations.
+   * @param error The calculated error.
+   * @param gradient The calculated gradient.
+   */
+  template<typename eT>
+  void Gradient(arma::Mat<eT>&& input,
+                arma::Mat<eT>&& error,
+                arma::Mat<eT>&& gradient);
+
+  //! Get the delta.
+  OutputDataType const& Delta() const { return delta; }
+  //! Modify the delta.
+  OutputDataType& Delta() { return delta; }
+
+  //! Get the gradient.
+  OutputDataType const& Gradient() const { return gradient; }
+  //! Modify the gradient.
+  OutputDataType& Gradient() { return gradient; }
+
+  /*//! Get the input parameter.
+  InputDataType const& InputParameter() const { return inputParameter; }
+  //! Modify the input parameter.
+  InputDataType& InputParameter() { return inputParameter; }*/
+
+  //! Return the model modules.
+  std::vector<LayerTypes<CustomLayers...> >& Model()
```

Comment:
I don't there is need for storing a model and `Add` given that you keep clearing the network. You can pass the single layer directly through the constructor.

Reply:
I thought it would be great to maintain consistency with the API. Also when we pass through the constructor we cannot change the layer inside the `wieghtNorm` wrapper.

Reply:
I believe you can create a getter and setter method for changing the layer. When you use the `Add` method you are not adding anything here which contradicts the purpose.
There is definitely no model or network that you are storing here just a single layer. 
Let me know what you think.

Reply:
Yeah, you are correct. I was earlier thinking about passing from constructor. But later on thought that we may add more than one layers to normalize. But yes I think it would be great to have only one layer.  So I am thinking to go with constructor now. Also I forgot about the getter and setter method as well.
Thanks for looking into it. I will make the changes soon.

Reply:
> But later on thought that we may add more than one layers to normalize. But yes I think it would be great to have only one layer. 

I think this may still be possible using `Sequential` and passing that to the `WeightNorm` constructor.

Reply:
No, Every layer there should be a different `vector and scalar` parameter. So that's not possible using sequential.